runtime/taskfirecrackeropts.go

```diff
@@ -0,0 +1,161 @@
+
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	firecracker "github.com/firecracker-microvm/firecracker-go-sdk"
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+	"github.com/pkg/errors"
+)
+
+// overrideVMConfigFromTaskOpts overrides the firecracker VM config for the
+// task with values set in the protobuf message.
+func overrideVMConfigFromTaskOpts(
+	cfg firecracker.Config,
+	vmConfig *proto.FirecrackerConfig,
+	drivesBuilder firecracker.DrivesBuilder,
+) (firecracker.Config, firecracker.DrivesBuilder, error) {
+	if vmConfig == nil {
+		return cfg, drivesBuilder, nil
+	}
+	var err error
+	if vmConfig.KernelImagePath != "" {
+		// If kernel image path is set, override it.
+		cfg.KernelImagePath = vmConfig.KernelImagePath
+	}
+	if vmConfig.KernelArgs != "" {
+		// If kernel args are set, override it.
+		cfg.KernelArgs = vmConfig.KernelArgs
+	}
+	// Attach network interface specified in the config.
+	if len(vmConfig.NetworkInterfaces) > 0 {
+		nwIfaces := make([]firecracker.NetworkInterface, len(vmConfig.NetworkInterfaces))
+		for i, nw := range vmConfig.NetworkInterfaces {
+			nwIfaces[i] = networkConfigFromProto(nw)
+		}
+		cfg.NetworkInterfaces = nwIfaces
+	}
+	if vmConfig.MachineCfg != nil {
+		cfg.MachineCfg, err = machineConfigFromProto(vmConfig.MachineCfg)
+		if err != nil {
+			return firecracker.Config{}, drivesBuilder, err
+		}
+	}
+	// Overwrite the root drive config if specified.
+	if vmConfig.RootDrive != nil {
+		drivesBuilder = drivesBuilderFromProto(vmConfig.RootDrive, drivesBuilder, true)
+	}
+	// Add additional drives if specified.
+	if len(vmConfig.AdditionalDrives) > 0 {
+		for _, drive := range vmConfig.AdditionalDrives {
+			drivesBuilder = drivesBuilderFromProto(drive, drivesBuilder, false)
+		}
+	}
+
+	return cfg, drivesBuilder, nil
+}
+
+// networkConfigFromProto creates a firecracker NetworkInterface object from
+// the protobuf FirecrackerNetworkInterface message.
+func networkConfigFromProto(nwIface *proto.FirecrackerNetworkInterface) firecracker.NetworkInterface {
+	result := firecracker.NetworkInterface{
+		MacAddress:  nwIface.MacAddress,
+		HostDevName: nwIface.HostDevName,
+		AllowMMDS:   nwIface.AllowMMDS,
+	}
+	if nwIface.InRateLimiter != nil {
+		result.InRateLimiter = rateLimiterFromProto(nwIface.InRateLimiter)
+	}
+	if nwIface.OutRateLimiter != nil {
+		result.OutRateLimiter = rateLimiterFromProto(nwIface.OutRateLimiter)
+	}
+
+	return result
+}
+
+// machineConfigFromProto creates a firecracker MachineConfiguration object
+// from the protobuf FirecrackerMachineConfiguration message.
+func machineConfigFromProto(mCfg *proto.FirecrackerMachineConfiguration) (models.MachineConfiguration, error) {
+	result := models.MachineConfiguration{}
+	if mCfg.MemSizeMib > 0 && mCfg.VcpuCount == 0 {
+		// Only memory is being overridden.
+		return result, errors.New("both vm memory and vcpu count need to be overridden")
+	}
+	if mCfg.VcpuCount > 0 && mCfg.MemSizeMib == 0 {
```

Comment:
nit: Can these be just one if `MemSizeMib == 0 || VcpuCount == 0` ?

Reply:
So, here we're verifying that if `vcpuCount` has been overridden, `memSizeMib` is overridden as well. Technically, you can have a machine config where you override neither. `MemSizeMib == 0 || VcpuCount == 0` would not work in that scenario.